### ISSUE TITLE
Rancher server mysql binlogs not rotating

### ISSUE THREAD

**Issue by flaccid** [rancher/rancher#2136]
```
root@c6e2d657f6f4:/var/log/mysql# ls -l
total 3791324
-rw-rw---- 1 mysql adm        8192 Sep 24 22:53 error.log
-rw-rw---- 1 mysql mysql 104857887 Sep  8 00:40 mysql-bin.000001
-rw-rw---- 1 mysql mysql 104858455 Sep  8 01:09 mysql-bin.000002
-rw-rw---- 1 mysql mysql 104857716 Sep  8 01:37 mysql-bin.000003
-rw-rw---- 1 mysql mysql 104858438 Sep  8 02:06 mysql-bin.000004
-rw-rw---- 1 mysql mysql 104875291 Sep  8 02:35 mysql-bin.000005
-rw-rw---- 1 mysql mysql 104858039 Sep  8 03:03 mysql-bin.000006
-rw-rw---- 1 mysql mysql 104857954 Sep  8 03:32 mysql-bin.000007
-rw-rw---- 1 mysql mysql 104857991 Sep  8 09:34 mysql-bin.000008
-rw-rw---- 1 mysql mysql 104859558 Sep  8 19:28 mysql-bin.000009
-rw-rw---- 1 mysql mysql 104904596 Sep  9 05:22 mysql-bin.000010
-rw-rw---- 1 mysql mysql 104868804 Sep  9 15:15 mysql-bin.000011
-rw-rw---- 1 mysql mysql 104880603 Sep  9 21:38 mysql-bin.000012
-rw-rw---- 1 mysql mysql 104860223 Sep 10 03:02 mysql-bin.000013
-rw-rw---- 1 mysql mysql 104881796 Sep 10 08:26 mysql-bin.000014
-rw-rw---- 1 mysql mysql 104858180 Sep 10 13:49 mysql-bin.000015
-rw-rw---- 1 mysql mysql 104865289 Sep 10 19:11 mysql-bin.000016
-rw-rw---- 1 mysql mysql 104864805 Sep 11 00:29 mysql-bin.000017
-rw-rw---- 1 mysql mysql 104905184 Sep 11 05:53 mysql-bin.000018
-rw-rw---- 1 mysql mysql 104864282 Sep 11 11:14 mysql-bin.000019
-rw-rw---- 1 mysql mysql 104874699 Sep 11 16:35 mysql-bin.000020
-rw-rw---- 1 mysql mysql 104857733 Sep 11 21:56 mysql-bin.000021
-rw-rw---- 1 mysql mysql 104857695 Sep 12 03:18 mysql-bin.000022
-rw-rw---- 1 mysql mysql 104858137 Sep 12 08:38 mysql-bin.000023
-rw-rw---- 1 mysql mysql 104879241 Sep 12 13:56 mysql-bin.000024
-rw-rw---- 1 mysql mysql 104863479 Sep 12 19:15 mysql-bin.000025
-rw-rw---- 1 mysql mysql 104862035 Sep 13 00:35 mysql-bin.000026
-rw-rw---- 1 mysql mysql 104858969 Sep 13 05:57 mysql-bin.000027
-rw-rw---- 1 mysql mysql 104871309 Sep 13 11:18 mysql-bin.000028
-rw-rw---- 1 mysql mysql 104860355 Sep 13 16:39 mysql-bin.000029
-rw-rw---- 1 mysql mysql 104860971 Sep 13 21:57 mysql-bin.000030
-rw-rw---- 1 mysql mysql 104862651 Sep 14 03:19 mysql-bin.000031
-rw-rw---- 1 mysql mysql 104860110 Sep 14 08:40 mysql-bin.000032
-rw-rw---- 1 mysql mysql 104857790 Sep 14 14:00 mysql-bin.000033
-rw-rw---- 1 mysql mysql 104861980 Sep 14 19:21 mysql-bin.000034
-rw-rw---- 1 mysql mysql 104866545 Sep 15 00:42 mysql-bin.000035
-rw-rw---- 1 mysql mysql 104858240 Sep 15 06:04 mysql-bin.000036
-rw-rw---- 1 mysql mysql 104858661 Sep 16 00:36 mysql-bin.000037
-rw-rw---- 1 mysql mysql   2027520 Sep 24 22:56 mysql-bin.000038
-rw-rw---- 1 mysql mysql      1216 Sep 16 00:36 mysql-bin.index
root@c6e2d657f6f4:/var/log/mysql# df -h
Filesystem                                              Size  Used Avail Use% Mounted on
none                                                    7.8G  7.7G     0 100% /
tmpfs                                                  1000M     0 1000M   0% /dev
shm                                                      64M     0   64M   0% /dev/shm
tmpfs                                                  1000M     0 1000M   0% /sys/fs/cgroup
/dev/disk/by-uuid/68da3afb-03e7-4e80-b493-5fa5eb34eedc  7.8G  7.7G     0 100% /etc/hosts
tmpfs                                                  1000M     0 1000M   0% /proc/kcore
tmpfs                                                  1000M     0 1000M   0% /proc/timer_stats
root@c6e2d657f6f4:/var/log/mysql# grep log /etc/mysql/my.cnf
# Be aware that this log type is a performance killer.
# As of 5.1 you can enable the log at runtime!
#general_log_file        = /var/log/mysql/mysql.log
#general_log             = 1
# Error log - should be very few entries.
log_error = /var/log/mysql/error.log
#log_slow_queries	= /var/log/mysql/mysql-slow.log
#log-queries-not-using-indexes
# The following can be used as easy to replay backup logs or for replication.
log_bin			= /var/log/mysql/mysql-bin.log
sync_binlog           = 1
innodb_flush_log_at_trx_commit = 1
expire_logs_days	= 10
max_binlog_size         = 100M
#binlog_do_db		= include_database_name
#binlog_ignore_db	= include_database_name
```

It seems the expiration and purging isn't working or perhaps there needs more configuration to ensure purging.

This is just a basic non-HA setup, running rancher in a single container on a 10GB host. It filled up and this is the 2nd time I've observed it happen. Can we look at how this can be prevented and keep the volume to a reasonable size?

**Comment by flaccid** [comment]
A little unclear in the mysql documentation if automatic flush will occur periodically or if a flush action needs to be manually actioned e.g. a mysql command or restarting the service.

**Comment by vincent99** [comment]
I *think* it only purges old files on startup or when creating a new log file.  So `.00038` was created on the 16th (just after `.00037` stopped updating) and at that time there were none more than 10 days old.  Since then there has been no rollover to a new file (and presumably no restart), so the ones that have become > 10 days old haven't been deleted.

You can connect to it and run something like `PURGE BINARY LOGS BEFORE DATE_SUB(NOW( ), INTERVAL 10 DAY);` to delete ones older than 10 days (...or change it to 1 day).

Unfortunately the combinations of settings (max size per file + max days) doesn't allow you to put an upper bound on the space taken.  We could turn off binlogs, but that makes it harder to replicate out as a migration strategy.  Maybe reduce the number of days to 1 or 2 @cloudnautique?

**Comment by aruneli** [comment]
@cloudnautique The issue is not fixed. In one of my setups, I still noticed this issue. So re-opening this. I will send the setup details privately.

**Comment by aruneli** [comment]
Referencing #2232

**Comment by cloudnautique** [comment]
@aruneli are you running v0.41.0-rc1? I thought the fix was in v0.40.0 but it did not make it in. It is in v0.41.0-rc1. You should see 2 day rotation. Please verify.

**Comment by aruneli** [comment]
Verified in 0.41.0 that mysql-bin files are purged.
